components/api/src/main/java/com/hotels/styx/api/StyxCoreObservable.java

```diff
@@ -62,31 +50,23 @@ public StyxCoreObservable(CompletionStage<T> future) {
         return new StyxCoreObservable<T>(Observable.error(cause));
     }
 
-    public <U> StyxObservable<U> map(Function<T, U> transformation) {
+    public <U> StyxObservable<U> map(Function<? super T, ? extends U> transformation) {
         return new StyxCoreObservable<>(delegate.map(transformation::apply));
     }
 
-    public <U> StyxObservable<U> flatMap(Function<T, StyxObservable<U>> transformation) {
+    public <U> StyxObservable<U> flatMap(Function<? super T, ? extends StyxObservable<? extends U>> transformation) {
```

Comment:
I am skeptical about `? extends StyxObservable<? extends U>` bit.

Due to some technical limitations, we don't intend `StyxObservable` to be extendable (although it is implemented by a package private subclass).

But the above annotation suggests that it is OK to provide a mapping function that provides `StyxObservable` subclasses.

Perhaps the function signature needs to be just: 

`Function<? super T, StyxObservable<? extends U>> transformation`

In fact this remark probably applies to all instances where we see this pattern repeating.

Reply:
I think StyxObservable is an interface, isn't it? :)  In this case, both signatures would accept any implementation of 'StyxObservable' (mmm, provided the declared type was for StyxObservable?) , but I think the one used by Kyle/rxJava would allow to pass  Function<StyxCoreObservable, SecondParam> whereas   your version only allows Function<StyxObservable, SecondParam>     .

Reply:
Upon making the suggested change to `flatMap`, a lot of the scala Specs stop compiling. It's not completely clear to me why, but I think there may be some significance to the new signature style (as seen in Rx) beyond what we currently realise.

Reply:
I have investigated a little bit an the problem with `Function<? super T, StyxObservable<? extends U>>` is that it will only accept a function whose return type is **verbatim** `StyxObservable<? extends U>`.

This means that a function that returns `StyxObservable<B>` where `B` is a subtype of `U` will not be allowed.

For this reason we need the `Function<? super T, ? extends StyxObservable<? extends U>>`.

Reply:
Ah, of course,   `<StyxObservable<Integer>>` is not compatible with `<StyxObservable<? extends Number>>` due to the type invariance of the first parametrized type. That´s what happens (to me) when reviewing code outside the IDE...